### ISSUE TITLE
Multiple CPU processes using same GPU model for inference

### ISSUE THREAD

**Issue by djdookie** [pytorch/pytorch#16943]
Hi guys,

under Windows, how can I use multiple processes, spawned by multiprocessing use the same GPU model for prediction?

I read this:
https://pytorch.org/docs/stable/multiprocessing.html#module-torch.multiprocessing
https://pytorch.org/docs/stable/notes/multiprocessing.html

According to this it seems possible: https://github.com/pytorch/pytorch/issues/3871
but under Windows?

According to this it seems it's not possible unter windows:
https://pytorch.org/docs/stable/notes/windows.html#cuda-ipc-operations

I only want to predict values from the same trained model from multiple spawned CPU processes, but currently I can't make it work.

I get this error:
THCudaCheck FAIL file=c:\a\w\1\s\tmp_conda_3.6_091443\conda\conda-bld\pytorch_1544087948354\work\torch\csrc\generic\StorageSharing.cpp line=232 error=71 : operation not supported

Can I make this somehow work under windows?

**Comment by ailzhang** [comment]
IIUC you want multiprocessing to work on GPU & windows right? 

CUDA IPC sharing was not supported on windows, looks like it's [supported now](https://github.com/NVIDIA/cuda-samples/commit/32f0fc6111b7e752b57effa702dfad585fa1ae51) @ngimel  Could you help confirm? Thanks!

**Comment by ngimel** [comment]
Yes, IPC is now supported on windows, see, e.g. cuda 9.2 release notes https://developer.download.nvidia.com/compute/cuda/9.2/Prod/docs/sidebar/CUDA_Toolkit_Release_Notes.pdf

**Comment by ailzhang** [comment]
cc: @ezyang  @VitalyFedyunin @peterjc123  for more visibility.

**Comment by peterjc123** [comment]
Do we need some additional changes to make it work on Windows? Because I see it's failing on this check `THCudaCheck(cudaIpcGetMemHandle(&handle, base_ptr));`. What's more, we have to switch our cuda to 10.0 in pytorch ci to test this feature. @yf225

**Comment by djdookie** [comment]
Thanks for your replies!
I am still using CUDA 9.0.
So if I install CUDA 9.2 or higher, it should work? Can I also use CUDA 10? Does pytorch support it?

In the code, how do I realise it?
Just create one model and give it into my multiprocessing.Pool as parameter should work then?

**Comment by sidazhang** [comment]
@ngimel as a side question, in general, does it even make sense to have multiprocessing workload on a GPU?

**Comment by VitalyFedyunin** [comment]
TODO: Add Windows GPU Multiprocessing tests.
@[djdookie] you might need completion of the #16854, as now GPU Multiprocessing requires very specific scopes management.

**Comment by djdookie** [comment]
Please don't get me wrong.
I don't want to do multiprocessing on GPU.
I want to do access **one single GPU model from multiple CPU processes** only for prediction under windows and python 3.6.

**Comment by VitalyFedyunin** [comment]
@djdookie apparently you are getting error inside of "static PyObject * THPStorage_(shareCuda)(THPStorage *self)" function. Which is considered interprocess communication.

You might want to wait #16854 to be finished or update cuda plus ensure that your model never get our of scope in the 'main' process.

**Comment by djdookie** [comment]
> 
> 
> @djdookie apparently you are getting error inside of "static PyObject * THPStorage_(shareCuda)(THPStorage *self)" function. Which is considered interprocess communication.
> 
> You might want to wait #16854 to be finished or update cuda plus ensure that your model never get our of scope in the 'main' process.

What do you mean with out of scope?

**Comment by ailzhang** [comment]
@djdookie The `out of scope` issue @VitalyFedyunin mentioned is explained here https://github.com/pytorch/pytorch/pull/16345/files
Are you trying to load a model trained on GPU, and then do inference on CPU multiprocessing?
In that case you can just map all the weights from gpu to cpu. using `map_location` in `torch.load`. Does that answer your question?

**Comment by djdookie** [comment]
Ok.

> Are you trying to load a model trained on GPU, and then do inference on CPU multiprocessing?

Yes, exactly this. Trained a model on GPU, then inferencing on CPU(s) by multiple processes.

I am doing this, which should be correct and not the problem I am facing:

```
    map_location = None if args.cuda else 'cpu'
    checkpoint = torch.load(filepath, map_location=map_location)
```

https://github.com/djdookie/alphastone/blob/f261dd01d3c1b94b5da582cfe2d2a367e47810a4/alphabot/NNet.py#L149

So, I will try CUDA 9.2 or 10 soon.

**Comment by ailzhang** [comment]
@djdookie If all your weights are loaded into cpu, I don't see why you got the error you posted in the description. Since that is only called when you try to get CUDAIpcmemhandle which only happens in multiprocessing on GPUs. 
And in this case switching to CUDA9.2 or 10 probably won't help in your case. 
I would check why on which part of your inference caused the error instead. If you could compose a minimal repro script, we can help on that as well.

**Comment by djdookie** [comment]
@ailzhang The above code is used when I don't use CUDA and have the model only on CPUs, which is what I am doing until now. But I am trying to switch to loading the model in CUDA/GPU and doing the rest on CPU. Can I use map_location=CPU when the model for inference is on GPU?

Sorry I am pretty new to pytorch. ;)

Small executable repro isn't easily possible, because I have a backing game-engine etc.
But you could have a quick look at my code at github if you want.

Quick explanation:

Main method is here at the very end instancing game engine and neural network wrapper:
https://github.com/djdookie/alphastone/blob/master/remote/selfplay_client.py
```
with Pool(self.args.numThreads) as pool:
    for result in list(tqdm(pool.imap(self.executeEpisode, range(self.args.numEps)), total=self.args.numEps, desc='Self-play matches')):
        iterationTrainExamples += result
```
Here I instantiate Coach() with the gameengine and neural network model (wrapper) as parameter.
Coach.learn() instanciates my MCTS class (again with both as parameters) and spawns multiple processes that should all use the same GPU-model if I set 'cuda': False in my wrapper class NNet.py (link is below)

Neural network wrapper class:
https://github.com/djdookie/alphastone/blob/master/remote/NNet.py

Deep residual network I am using:
https://github.com/djdookie/alphastone/blob/master/remote/alphanet12.py

**Comment by ailzhang** [comment]
@djdookie 
> But I am trying to switch to loading the model in CUDA/GPU and doing the rest on CPU.

If I understand your question correctly, your model weights are in GPU and you want to do the rest on CPU. `map_location` is designed exactly for this, by mapping it to cpu, you can get weights (trained on GPU) loaded to CPU and start everything else with CPU.

`torch.load('tensors.pt', map_location=torch.device('cpu'))`

The confusion we had was, the error message you posted was indeed trying to share a cuda storage, which is not what you want to do I assume.

**Comment by djdookie** [comment]
Ok now I am on CUDA 10, python 3.7, pytorch 1.0.1, numpy 1.15.4 with mkl.

If I do the following:
```
state = torch.FloatTensor(state.astype(np.float64)).unsqueeze(0).unsqueeze(0)
state = state.contiguous().cuda()
```
and
`torch.load(filepath, torch.device('cpu'))`

I am still getting this error:
`THCudaCheck FAIL file=c:\a\w\1\s\tmp_conda_3.7_183942\conda\conda-bld\pytorch_1549564939537\work\torch\csrc\generic\StorageSharing.cpp line=232 error=71 : operation not supported`

Why is this? I don't get it...

If I map weights to CPU I can do predictions on GPU?

**Comment by djdookie** [comment]
@ailzhang @VitalyFedyunin 
In my case model.predict() takes
- 0.01 sec on my Geforce GTX 1080.
- 0.35 sec on my Intel i7 4770K. (thats 35x slower on CPU compared with my GPU)

I can easily run parallel processes in the second case (only on CPU, without GPU/CUDA).
But I want to do the first thing n parallel -> call model.predict() on GPU.

How do I do that with pytorch?

**Comment by VitalyFedyunin** [comment]
Related to the `StorageSharing.cpp line=232 error=71 : operation not supported` error.
As per @ngimel "IPC functionality is supported on win, but only in TCC mode https://docs.nvidia.com/gameworks/content/developertools/desktop/nsight/tesla_compute_cluster.htm". 

Besides that I see no problems to train on GPU and do inference on CPU.

**Comment by djdookie** [comment]
@VitalyFedyunin I want to do inference on GPU, but in multiple parallel processes.

**Comment by VitalyFedyunin** [comment]
Don't panic. Let's tackle problems one by one.
"THCudaCheck FAIL file=c:\a\w\1\s\tmp_conda_3.6_091443\conda\conda-bld\pytorch_1544087948354\work\torch\csrc\generic\StorageSharing.cpp line=232 error=71 : operation not supported" clearly because of the Windows without TCC mode. As soon as you fix it, we can look at your code and help you to do proper synchronization (if necessary).

**Comment by djdookie** [comment]
@VitalyFedyunin I can't use TCC because it's my main GPU with an active display.

```
C:\Program Files\NVIDIA Corporation\NVSMI>nvidia-smi -g 0 -dm 1
Unable to set driver model for GPU 00000000:01:00.0: TCC can't be enabled for device with active display.
Treating as warning and moving on.
All done.
```

I believe I want to do something like this, just for prediction not training:
https://github.com/pytorch/examples/blob/master/mnist_hogwild/main.py

**Comment by djdookie** [comment]
Just sawthat the mnist_hogwild_example is without cuda. Would it be possible to use cuda here with shared_memory for predictions?

**Comment by djdookie** [comment]
Ok I changed the code, trying to implement CUDA processing and get the same error.
Now you have short reproducible code:

https://github.com/djdookie/examples/tree/master/mnist_hogwild

**Comment by djdookie** [comment]
Interestingly, if I just start individual processes using the subprocess module, it works super performant:
(no great prediction time loss)
```
import subprocess
import time

for i in range(8):
    subprocess.Popen('python selfplay_client.py')
    time.sleep(1)
```

I wonder why this isn't that simple in pytorch...

**Comment by igiloh** [comment]
@djdookie I've been solving a similar problem lately (on linux) - predicting from several different CPU processes, using the same model on the GPU. I've ended up taking a different approach, which might be helpful for you.

My use case was a gateway that handles image prediction requests. Each request requires I/O (reading the image), preprocessing (on the CPU), prediction (GPU), post-processing (CPU again) and I/O out (writing outputs).  

I've tried with the two trivial solutions:
1. Have a single process load a GPU model, then share it with other processes using `model.share_memory()`.
2. Have all the processes marshall their inputs to the GPU, then share these with the main "prediction" process.
None of these worked well - as it seems that each process handled its own CUDA cache separately, which very quickly escalated to a GPU memory overrun.

My (working) solution was to adopt a "multi-producer single consumer" approach - by memory sharing **CPU** tensors.  
A single process (let's call it the "GPU worker") loads a torch model on the GPU. All other "worker processes" are handling the I/O, pre/post processing etc. When a worker wants to get a model prediction - memory shares its (CPU) tensor,  and puts it on a single queue.  
The GPU worker keeps polling the queue, marshalls the input tensor to the GPU (`.to(device)`), runs the prediction and marshalls the output back to the CPU. It then communicate the result back to the originating worker - using a separate queue per worker.  

This approach can be easily used also for batch predictions - the GPU worker can aggregate a few tensors from the queue to a single batch, do a single prediction on all of them, then split the results back to the workers (I haven't implemented that part yet...).

I hope this is somehow helpful for you. From my own personal experience this approach was far more stable than CUDA tensors sharing - yet the performance benefits were enormous .

**Comment by hfarhidzadeh** [comment]
I am on the same page. Trained model on multiple GPUs, trying to do inference on multiple CPUs . The problem I have is in higher layers of model, I am getting cache dump in abc.py __instancecheck__  for _abc_negative_cache!

**Comment by VitalyFedyunin** [comment]
@hfarhidzadeh can you please attach code/pseudo-code of your solution as well as error log.

**Comment by hfarhidzadeh** [comment]
@VitalyFedyunin I am working on this code:
[https://github.com/mattmacy/vnet.pytorch/blob/master/vnet.py](url)

trained model on 3 P4000 GPUs and tested on single M4000 GPU works fine.
The error comes from : 
Line 178 -> out = self.up_tr32(out, out16) -> Line 106: out = self.ops(xcat)
on inference side on CPU.
First I tried on 16 GB RAM which seems is not sufficient (error: Need 31GB RAM, buy new one!)

Then I tested on machine with 128 GB RAM and it crashed without showing any error. By tracing, I figured out problem is from abc.py:

 ` def __instancecheck__(cls, instance):
        """Override for isinstance(instance, cls)."""
        # Inline the cache checking
        subclass = instance.__class__
        if subclass in cls._abc_cache:
            return True
        subtype = type(instance)
        if subtype is subclass:
            if (cls._abc_negative_cache_version ==
                ABCMeta._abc_invalidation_counter and
                subclass in cls._abc_negative_cache):
                return False
            # Fall back to the subclass check.
            return cls.__subclasscheck__(subclass)
        return any(cls.__subclasscheck__(c) for c in {subclass, subtype})`
 
Thank you. Let me know if it is not still clear.

**Comment by stefanlopez** [comment]
@VitalyFedyunin do you know if IPC functionality will ever be supported on windows without TCC mode? I would really like to run PyTorch in parallel on windows with my RTX 2070, but it doesn't seem that feasible right now.

**Comment by quantumlicht** [comment]
@igiloh, 
In reference to https://github.com/pytorch/pytorch/issues/16943#issuecomment-463118535
Do you have a snippet of code you could share? I am trying to do something very similar for batching inference requests.

**Comment by cbasavaraj** [comment]
@igiloh Looks like you are performing inference sequentially? It's nice to economise on GPU memory usage, but at the same time we want to have low latency right? The way I see it, the model weights need only be read, so multiple processes (maybe not threads due to python GIL) should be able to access and read the weights to perform inference. But in practice, haven't been able to get it work at all. I think this is like a holy grail for deployment, more than TorchScript or TorchServe.

**Comment by tobiaswuerth** [comment]
no solution for windows anywhere yet it seems :/

**Comment by taras-janea** [comment]
@djdookie, could you please try one of the approaches suggested in the [comment](https://github.com/pytorch/pytorch/issues/16943#issuecomment-463118535) using the latest PyTorch version?

**Comment by taras-janea** [comment]
Closing this issue due to inactivity. If you have any related concerns, please don't hesitate to open a new issue.
